### PR TITLE
feat(query): support array aggregate functions

### PR DESCRIPTION
**Pull request by b41sh**
I hereby agree to the terms of the CLA available at: https://databend.rs/dev/policies/cla/

## Summary

support array aggregate functions
```
array_sum
array_avg
array_count
array_max
array_min
```

part of #7931

### CODE REVIEW

File: src/query/functions/src/scalars/array.rs [hunk 1]
```diff
@@ -606,4 +613,381 @@ pub fn register(registry: &mut FunctionRegistry) {
             }
         }),
     );
+
+    for left in ALL_SIGNED_INTEGER_TYPES {
+        with_number_mapped_type!(|NUM_TYPE| match left {
+            NumberDataType::NUM_TYPE => {
```
Comment (sundy-li) [inline src/query/functions/src/scalars/array.rs:619]: can we use `register_factory` and agg combinators?

The DataType could be evaluated as `agg.input_type()`.
Thus we could make the code path cleaner.

Comment (b41sh) [reply src/query/functions/src/scalars/array.rs:619]: The `return_type` is related to the type of the aggregate operation, and the type of the operation needs to be obtained by the second argument, so the `return_type` cannot be known in `FunctionSignature`.

File: src/query/functions/src/scalars/array.rs [hunk 2]
```diff
@@ -614,380 +613,142 @@ pub fn register(registry: &mut FunctionRegistry) {
...
-                            output.push(val);
-                        }
-                        Err(err) => {
-                            ctx.set_error(output.len(), err.to_string());
-                        }
-                    }
-                } else {
-                    output.push_null();
-                }
+    registry.register_function_factory("array_max", |_, args_type| {
```
Comment (sundy-li) [inline src/query/functions/src/scalars/array.rs:727]: can we use for loop to register these array_agg functons ?

**Review by sundy-li — APPROVED** [review]

File: src/query/functions/src/scalars/array.rs [hunk 3]
```diff
@@ -61,12 +64,25 @@ use itertools::Itertools;
 use siphasher::sip128::Hasher128;
 use siphasher::sip128::SipHasher24;
 
+use crate::aggregates::eval_aggr;
+use crate::AggregateFunctionFactory;
+
+const ARRAY_AGGREGATE_FUNCTIONS: &[(&str, &str); 5] = &[
+    ("array_avg", "avg"),
```
Comment (sundy-li) [inline src/query/functions/src/scalars/array.rs:71]: const ARRAY_AGGREGATE_FUNCTIONS =  ["avg", "min", "max", ...];

and use `concat!("arr_", agg)`

**Pull request merged**